File: src/skillbank/emitters/canonical.py
"""Canonical emitter — SkillIR → canonical SKILL.md(写回中央仓)。

零损耗实现关键:
- body bytes 直接拼到 frontmatter 之后, 不 decode/encode/normalize
- frontmatter: 优先字段级透传(字节稳定)——parser 保留原始 frontmatter 字节,
  未变更的字段原字节保留, 只重写真正改动(新增/删除/修改)的字段。
  否则 safe_dump 全量重建会丢引号、改字段顺序(git diff 产生无意义噪音)。
  仅当 IR 无原始 frontmatter(如 import 新建)时, 才 safe_dump 全量生成。

roundtrip zero-loss: parse->IR0 -> emit_canonical -> parse -> IR1
    assert IR1.body == IR0.body  (字节完全等值)
    assert frontmatter 未变字段字节不变(字段级透传)
"""

from __future__ import annotations

from pathlib import Path

import yaml

from skillbank.emitters.base import edit_frontmatter_fields
from skillbank.ir import SkillIR

__all__ = ["emit_canonical"]


def emit_canonical(ir: SkillIR, target_path: Path) -> None:
    """把 IR 写成 canonical SKILL.md。

    有原始 frontmatter(fm_raw/fm_orig)时走字段级透传(字节稳定);
    否则(如 import 新建)safe_dump 全量生成。
    Args:
        ir: SkillIR(必填 name/description)
        target_path: 写入路径(父目录需存在; 不自动建目录)
    """
    if ir.fm_raw is not None and ir.fm_orig is not None:
        raw_text = ir.fm_raw.decode("utf-8")
        edited = edit_frontmatter_fields(raw_text, ir.fm_orig, ir.to_frontmatter_dict())
        fm_bytes = edited.rstrip("\n").encode("utf-8") + b"\n"
    else:
        fm = ir.to_frontmatter_dict()
        fm_bytes = yaml.safe_dump(
            fm, sort_keys=False, allow_unicode=True,
            default_flow_style=False, width=10_000,
        ).encode("utf-8")

    out_bytes = b"---\n" + fm_bytes + b"---\n" + ir.body
    Path(target_path).write_bytes(out_bytes)

File: internal/emit/canonical.go
// canonical emitter — SkillIR → canonical SKILL.md(写回中央仓)。
//
// 零损耗实现关键(对应 Python emitters/canonical.py):
//   - body bytes 直接拼到 frontmatter 之后, 不 decode/encode/normalize
//   - 有原始 frontmatter → 字段级透传(未变更字段原字节保留)
//   - 无原始 frontmatter(如 import 新建)→ 全量 dump
//
// roundtrip zero-loss: parse→IR0 → EmitCanonical → parse → IR1,
// IR1.Body == IR0.Body 字节等值。
package emit

import (
	"os"
	"strings"

	"github.com/ite-rate/skillbank/internal/ir"
)

// EmitCanonical — 把 IR 写成 canonical SKILL.md。
// 父目录需存在, 不自动建目录(与 Python 一致, add/importer 层负责建)。
func EmitCanonical(in *ir.SkillIR, targetPath string) error {
	fmBytes, err := canonicalFrontmatterBytes(in)
	if err != nil {
		return err
	}
	out := append([]byte("---\n"), fmBytes...)
	out = append(out, []byte("---\n")...)
	out = append(out, in.Body...)
	return os.WriteFile(targetPath, out, 0o644)
}

// CanonicalSkillMDBytes — 不落盘版本(add/import 内存预览、测试用)。
func CanonicalSkillMDBytes(in *ir.SkillIR) ([]byte, error) {
	fmBytes, err := canonicalFrontmatterBytes(in)
	if err != nil {
		return nil, err
	}
	out := append([]byte("---\n"), fmBytes...)
	out = append(out, []byte("---\n")...)
	return append(out, in.Body...), nil
}

func canonicalFrontmatterBytes(in *ir.SkillIR) ([]byte, error) {
	fm := in.ToFrontmatterDict()
	if in.FMRaw != nil && in.FMOrig != nil {
		edited, err := EditFrontmatterFields(string(in.FMRaw), in.FMOrig, fm,
			ir.CanonicalFieldOrder)
		if err != nil {
			return nil, err
		}
		return []byte(strings.TrimRight(edited, "\n") + "\n"), nil
	}
	return DumpMap(fm, ir.CanonicalFieldOrder)
}